src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/PoolingDataSource.cs

```diff
@@ -424,13 +428,46 @@ internal void PruneIdleConnections()
             throw new NotImplementedException();
         }
 
+        private void QueueWarmupTask()
+        {
+            // Make sure we're at quota by posting a callback to the threadpool.
+            ThreadPool.QueueUserWorkItem(_warmupCallback);
+        }
+
         /// <summary>
         /// Warms up the pool to bring it up to min pool size.
         /// </summary>
         /// <exception cref="NotImplementedException"></exception>
-        internal void WarmUp()
+        private async void WarmupCallback(object? state)
```

Comment:
It's generally a good idea to avoid async void methods whenever possible ([article](https://learn.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming#avoid-async-void)); if you have to use them, it's worth paying very close attention to what happens with exceptions - any exception that bubbles out of an async void method causes AppDomain.UnhandledException to get fired.

I'd recommend making this async Task and assigning the Task to some instance variable on the pool; your could even make sure that Task is complete when the pool is shuw down/disposed (pretty contrived edge case, but at least for completeness).

Otherwise if you keep async void, maybe consider wrapping the method body with a try/catch that doesn't bubble any exceptions out (but rather logs).

Reply:
I'm just going to push it over the fence and say "this should be `async Task` or `async ValueTask`" :)

Reply:
Since we're thinking about having this as a `Task`, then I suggest using `TaskCreationOptions.LongRunning`, so that the pool maintenance task doesn't sit in the normal thread pool (and won't generate/suffer from the effects of thread pool exhaustion.)

We also definitely want to be able to cancel the maintenance task using a CancellationToken - I think `SqlConnection.ClearAllPools()` will need to pull that lever.

Reply:
The way I'm reading the `TaskCreationOptions.LongRunning` option is that it makes sense if you're doing a lot of synchronous work on that Task. This warmup process and the broader pool maintenance tasks will all be async, so I believe it won't make sense to set that option.

Reply:
I got this working using a task stored in an instance variable. It also improved the unit testing as this task can be awaited directly to determine when warmup has completed.

Reply:
Yes, you're right - not sure how I missed that, sorry! I can see the CancellationToken is in place, so I'm happy with the methodology.